crates/m3/src/builder/witness.rs

```diff
@@ -37,51 +39,120 @@ use crate::builder::multi_iter::MultiIterator;
 /// gets converted into a multilinear extension index, which maintains references to the data
 /// allocated by the allocator, but does not need to maintain a reference to the constraint system,
 /// which can then be dropped.
-#[derive(Debug, Default, CopyGetters)]
+#[derive(Debug)]
 pub struct WitnessIndex<'cs, 'alloc, P = PackedType<OptimalUnderlier, B128>>
 where
 	P: PackedField,
 	P::Scalar: TowerField,
 {
-	pub tables: Vec<Option<TableWitnessIndex<'cs, 'alloc, P>>>,
+	allocator: &'alloc Bump,
+	/// Each entry is Left if the index hasn't been initialized & filled, and Right if it has.
+	tables: Vec<Either<&'cs Table<P::Scalar>, TableWitnessIndex<'cs, 'alloc, P>>>,
```

Comment:
Another thing I would consider here, is to simply have the WitnessIndex hold an immutable reference to the ConstraintSystem, and have this be a `Vec<Option<TableWitnessIndex<'cs, 'alloc, P>>`

Reply:
I initially had something like that, but prefer having a single vector instead of two vectors that must be the same length by invariant.